### PR TITLE
Client commands code improvements + ingame support

### PR DESCRIPTION
**Pull request by dydzio0614**
Command-line client commands now work from ingame console and got moved to separate file

### CODE REVIEW

**Comment by IvanSavenko** [comment]
(haven't looked in detail)
I suggest to use some syntax to differentiate chat & commands.
It is quite possible for player to write (for example) "save" in chat to well, ask another player in MP to save game.
But right now I assume this will trigger a command?

Proposition: only attempt to interpret as commands strings that start with '/'.
So, if string starts with '/' then send to commands processor everything sans '/'
Otherwise - process message as game chat

**Comment by dydzio0614** [comment]
what about "gosolo" pseudo-cheat from original H3

**Comment by IvanSavenko** [comment]
Personally I'd use it unconditionally for any command, meaning - also for gosolo

File: client/ClientCommandManager.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#ifndef VCMI_IOS
+class PlayerColor;
```
Comment (IvanSavenko) [inline client/ClientCommandManager.h:14]: Should in theory fix ios compile
```suggestion
VCMI_LIB_NAMESPACE_BEGIN
class PlayerColor;
VCMI_LIB_NAMESPACE_END
```

File: client/widgets/AdventureMapClasses.cpp [hunk 2]
```diff
@@ -1144,15 +1145,22 @@ void CInGameConsole::startEnteringText()
...
 		std::string txt = enteredText.substr(0, enteredText.size()-1);
 		LOCPLINT->cb->sendMessage(txt, LOCPLINT->getSelection());
 		previouslyEntered.push_back(txt);
+
+		if(txt.at(0) == '/')
+		{
+			//some commands like gosolo don't work when executed from GUI thread
+			boost::thread clientCommandThread(ClientCommandManager::processCommand, txt.substr(1), true);
+			clientCommandThread.detach();
+		}
```
Comment (IvanSavenko) [inline client/widgets/AdventureMapClasses.cpp:1163]: Proposition: move `LOCPLINT->cb->sendMessage` into else block here.
Meaning, don't send console commands as chat messages

File: client/ClientCommandManager.cpp [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+	}
+	else if(commandName == "controlai")
+	{
+		std::string colorName;
+		readed >> colorName;
+		boost::to_lower(colorName);
+
+		ClientCommandManager::handleControlAi(colorName);
+	}
+	// Check mantis issue 2292 for details
```
Comment (IvanSavenko) [inline client/ClientCommandManager.cpp:454]: Perhaps add "command no recognized" message or something like that?

File: client/ClientCommandManager.cpp [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+			logGlobal->error(commandMessage);
+			break;
+		default:
+			std::cout << commandMessage;
+			break;
+	}
+
+	if(currentCallFromIngameConsole)
+	{
+		if(LOCPLINT && LOCPLINT->cingconsole)
```
Comment (IvanSavenko) [inline client/ClientCommandManager.cpp:535]: Access to GUI objects from another thread requires lock.
I think this is correct way to lock, but please recheck
```suggestion
		boost::unique_lock<boost::recursive_mutex> un(*CPlayerInterface::pim);
		if(LOCPLINT && LOCPLINT->cingconsole)
```

**Review by IvanSavenko — APPROVED** [review]

File: client/CMakeLists.txt [hunk 5]
```diff
@@ -82,6 +82,7 @@ set(client_SRCS
 		NetPacksClient.cpp
 		NetPacksLobbyClient.cpp
 		SDLRWwrapper.cpp
+        ClientCommandManager.cpp
```
Comment (Nordsoft91) [inline client/CMakeLists.txt:85]: Please use tabs

File: lib/CConsoleHandler.h [hunk 6]
```diff
@@ -76,7 +76,7 @@ class DLL_LINKAGE CConsoleHandler
 #endif
 	}
 
-    std::function<void(const std::string &)> *cb; //function to be called when message is received
+    std::function<void(const std::string &, bool)> *cb; //function to be called when message is received
```
Comment (Nordsoft91) [inline lib/CConsoleHandler.h:79]: Please, add comment about meaning of bool argument

File: client/ClientCommandManager.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	static void giveTurn(const PlayerColor &color);
+	static void printInfoAboutInterfaceObject(const CIntObject *obj, int level);
+	static void printCommandMessage(const std::string &commandMessage, ELogLevel::ELogLevel messageType = ELogLevel::NOT_SET);
+	static void handleGoSolo();
+	static void handleControlAi(const std::string &colorName);
+
+public:
+	ClientCommandManager() = delete;
+	static void processCommand(const std::string &message, bool calledFromIngameConsole);
+};
```
Comment (Nordsoft91) [inline client/ClientCommandManager.h:31]: it doesn't manage something, it's event not a class but set of functions. So why not have set of functions wrapped by namespace?

File: client/Client.cpp [hunk 8]
```diff
@@ -761,6 +761,20 @@ void CClient::reinitScripting()
...
+void CClient::removeGUI()
+{
+	// CClient::endGame
+	GH.curInt = nullptr;
+	if(GH.topInt())
+		GH.topInt()->deactivate();
+	adventureInt = nullptr;
+	GH.listInt.clear();
+	GH.objsToBlit.clear();
+	GH.statusbar = nullptr;
```
Comment (Nordsoft91) [inline client/Client.cpp:773]: are curInt, adventureInt and statusBar shared_ptrs?
If so, please use reset() instead.
If not, isn't it a memory leak?

**Review by Nordsoft91 — CHANGES_REQUESTED** [review]

File: client/ClientCommandManager.h (hunk 7 above)
Comment (IvanSavenko) [reply client/ClientCommandManager.h:31]: It does use "private" visibility to hide internal stuff from outside calls.
Perhaps remove 'static' and use it as proper class instead?
As for name - Controller or Executor perhaps?

File: client/Client.cpp (hunk 8 above)
Comment (dydzio0614) [reply client/Client.cpp:773]: This code was moved from different file without alteration

Comment (IvanSavenko) [reply client/Client.cpp:773]: Not a memory leak - `operator =` will correctly release object if needed.
In fact, shared_ptr does not have a way to release object, so no memory leaks are possible.
(unique_ptr does have release(), but not shared_ptr)

File: client/ClientCommandManager.h (hunk 7 above)
Comment (dydzio0614) [reply client/ClientCommandManager.h:31]: so you recommend creating new instance for each usage?

Comment (IvanSavenko) [reply client/ClientCommandManager.h:31]: In my opinion - yes, but on stack, without call to new

Comment (dydzio0614) [reply client/ClientCommandManager.h:31]: Unsure about that yet, since this class is context independent and ability to freely create its object doesn't bring anything interesting to the table. The internal state isn't interesting for caller. Command can be executed at any arbitrary place in code so being static makes sense, from "domain driven design" point of view etc.

Comment (IvanSavenko) [reply client/ClientCommandManager.h:31]: Personally don't care either way, as long as it stays as "class" - it is weird to have a namespace with random functions that should not be called.
Nordsoft didn't liked this "static class" approach, so I suggested alternative.

Comment (dydzio0614) [reply client/ClientCommandManager.h:31]: well, in theory it could be singleton class, but does it matter? For example in C# you have Math class with bunch of static methods, Unity engine adds Random with another static methods (C# standard libraries also has Random but it's non static for some reason and often wrapped in some singleton so you don't create new instances over and over for the sake of good RNG seed)

Comment (IvanSavenko) [reply client/ClientCommandManager.h:31]: Keep in mind that C#
1) Does not have concept of "namespace with functions", all functions must be part of a class, so Nordsoft proposition is something that only exists in C++
2) I assume C# uses singleton in these cases to avoid (relatively) costly object creation call. In C++, creation on stack is extremely cheap operation. Especially for a state-less class.
3) Singleton is often viewed as an anti-pattern. Our VLC would be a great example of such.

By the way, your class is *not* stateless due to `currentCallFromIngameConsole` member. So yeah, I also suggest to change this, but into a proper class, with allocation on stack.

Comment (dydzio0614) [reply client/ClientCommandManager.h]: Done, though the currentCallFromIngameConsole is there for convenience and it is rewritten on every public function call - I just do not want to pass bool as argument everywhere within class

Comment (IvanSavenko) [reply client/ClientCommandManager.h]: Sure, my complain against `currentCallFromIngameConsole` was mostly about having it as static (which is now fixes)
While this is close to impossible with current design, if you somehow do end up with two instances of your class, you could end up with them accessing same variable since it was static.

Comment (dydzio0614) [reply client/ClientCommandManager.h]: would be still technically impossible under normal conditions for one client to type 2 messages from 2 sources at same time unless deliberately trying to break something

File: client/Client.cpp (hunk 8 above)
Comment (dydzio0614) [reply client/Client.cpp]: BTW... curInt and LOCPLINT are... not very smart pointers :laughing:

File: client/CMakeLists.txt [hunk 9]
```diff
@@ -168,6 +169,7 @@ set(client_HEADERS
 		mapHandler.h
 		resource.h
 		SDLRWwrapper.h
+        ClientCommandManager.h
```
Comment (Nordsoft91) [inline client/CMakeLists.txt:172]: I assumed that you will fix spaces in both places, but seems like need to leave comment separately :)

File: lib/CConsoleHandler.h [hunk 10]
```diff
@@ -75,8 +75,8 @@ class DLL_LINKAGE CConsoleHandler
 		funlockfile(stdout);
 #endif
 	}
-
-    std::function<void(const std::string &)> *cb; //function to be called when message is received
+	//function to be called when message is received - string: message, bool: whether call was made from in-game console
+    std::function<void(const std::string &, bool)> *cb;
```
Comment (Nordsoft91) [inline lib/CConsoleHandler.h:79]: Please, replace spaces with tab

**Review by Nordsoft91 — APPROVED** [review]
